pkg/kv/kvserver/rangefeed/bench_test.go

```diff
@@ -0,0 +1,211 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package rangefeed
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/isolation"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/util/future"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/stretchr/testify/require"
+)
+
+type benchmarkRangefeedOpts struct {
+	opType           opType
+	numRegistrations int
+	disableBudget    bool
+}
+
+type opType string
+
+const (
+	writeOpType    opType = "write"  // individual 1PC writes
+	commitOpType   opType = "commit" // intent + commit writes, 1 per txn
+	closedTSOpType opType = "closedts"
+)
+
+// BenchmarkRangefeed benchmarks the processor and registrations, by submitting
+// a set of events and waiting until they are all emitted.
+func BenchmarkRangefeed(b *testing.B) {
+	for _, opType := range []opType{writeOpType, commitOpType, closedTSOpType} {
+		for _, numRegistrations := range []int{1, 10, 100} {
+			name := fmt.Sprintf("opType=%s/numRegs=%d", opType, numRegistrations)
+			b.Run(name, func(b *testing.B) {
+				runBenchmarkRangefeed(b, benchmarkRangefeedOpts{
+					opType:           opType,
+					numRegistrations: numRegistrations,
+				})
+			})
+		}
+	}
+}
+
+// BenchmarkRangefeedBudget benchmarks the effect of enabling/disabling the
+// processor budget. It sets up a single processor and registration, and
+// processes a set of events.
+func BenchmarkRangefeedBudget(b *testing.B) {
+	for _, budget := range []bool{false, true} {
+		b.Run(fmt.Sprintf("budget=%t", budget), func(b *testing.B) {
+			runBenchmarkRangefeed(b, benchmarkRangefeedOpts{
+				opType:           writeOpType,
+				numRegistrations: 1,
+				disableBudget:    !budget,
+			})
+		})
+	}
+}
+
+// runBenchmarkRangefeed runs a rangefeed benchmark, emitting b.N events across
+// a rangefeed.
+func runBenchmarkRangefeed(b *testing.B, opts benchmarkRangefeedOpts) {
+	defer log.Scope(b).Close(b)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+
+	var budget *FeedBudget
+	if !opts.disableBudget {
+		budget = newTestBudget(math.MaxInt64)
+	}
+	span := roachpb.RSpan{Key: roachpb.RKey("a"), EndKey: roachpb.RKey("z")}
+
+	// Set up processor.
+	p := NewProcessor(Config{
+		AmbientContext:   log.MakeTestingAmbientContext(nil),
+		Clock:            hlc.NewClockForTesting(nil),
+		Metrics:          NewMetrics(),
+		Span:             span,
+		MemBudget:        budget,
+		EventChanCap:     b.N,
+		EventChanTimeout: time.Hour,
+		PushTxnsInterval: 0,
+		PushTxnsAge:      0,
```

Comment:
We can omit those, they were only explicit in tests because we were setting values if txn pusher was provided.